zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/GroupEntityAddedApplier.java

```diff
@@ -37,7 +37,10 @@ public void applyState(final long key, final GroupRecord value) {
     final var entityKey = value.getEntityKey();
     final var entityType = value.getEntityType();
     switch (entityType) {
-      case USER -> userState.addGroup(entityKey, groupKey);
+      case USER ->
+          userState
+              .getUser(entityKey)
+              .ifPresent(user -> userState.addGroup(user.getUsername(), groupKey));
```

Comment:
💭 The user check is redundant here, as we perform it in the processor. But I guess this is what you meant in the PR description, that first, we need to refactor the tenant/group/role states to work with IDs instead of keys.

And you need to retrieve the `username` somehow 🙂

Reply:
Exactly. It's not here to check if the user exist, it's to find the username 😄